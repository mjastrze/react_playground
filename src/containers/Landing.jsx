import React from 'react';
import { Link } from 'react-router-dom';

export const Landing = () => (
  <div className="landing">
    <h1>Movies</h1>
    <input type="text" placeholder="Search" />
    <Link alt="dsf" to="/search">
      or browse all
    </Link>
  </div>
);
