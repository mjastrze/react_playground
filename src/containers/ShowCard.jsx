import React from 'react';
import { shape, string } from 'prop-types';
import { Link } from 'react-router-dom';

import styled from 'styled-components';

// //styled components
const Wrapper = styled(Link)`
  width: 32%;
  border: 2px solid #333;
  border-radius: 4px;
  margin-bottom: 25px;
  padding-right: 10px;
  overflow: hidden;
  color: black;
  text-decoration: none;

  & > h1 {
    color: red;
  }
`;

const Image = styled.img`
  width: 46%;
  float: left;
  margin-right: 10px;
`;

export const ShowCard = ({
  show: { title, poster, year, description, imdbID },
}) => (
  <Wrapper to={`details/${imdbID}`}>
    <Image alt={`${title} Show Poster`} src={`/public/img/posters/${poster}`} />
    <div>
      <h3>{title}</h3>
      <h4>({year})</h4>
      <p>{description}</p>
    </div>
  </Wrapper>
);

ShowCard.propTypes = {
  show: shape({
    poster: string.isRequired,
    title: string.isRequired,
    year: string.isRequired,
    description: string.isRequired,
    imdbID: string.isRequired,
    trailer: string.isRequired,
  }).isRequired,
};
