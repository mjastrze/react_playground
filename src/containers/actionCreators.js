import { SET_SEARCH_TERM, ADD_API_DATA } from './actions';

export function setSearchTerm(searchTerm) {
  return { type: SET_SEARCH_TERM, payload: searchTerm };
}

export function resetSearchTerm() {
  return { type: SET_SEARCH_TERM, payload: '' };
}

export function addAPIData(apiData) {
  return { type: ADD_API_DATA, payload: apiData };
}

export function getAPIDetails(imdbID) {
  return (dispatch) => {
    fetch(`http://localhost:3000/${imdbID}`)
      .then((response) => response.json())
      .then((data) => dispatch(addAPIData(data)))
      .catch(console.error); // eslint-disable-line no-console
  };
}
