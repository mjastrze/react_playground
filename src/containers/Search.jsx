import React from 'react';
import data from '../data.json';
import { ShowCard } from './ShowCard';

// import style from './Search.module.css';

export class Search extends React.Component {
  state = {
    searchTerm: 'dupa',
  };

  handleSearchTermChange = (event) => {
    this.setState({
      searchTerm: event.target.value,
    });
  };

  render() {
    const { searchTerm } = this.state;
    return (
      <div className="search">
        <header>
          <input
            type="text"
            placeholder="Search"
            value={searchTerm}
            onChange={this.handleSearchTermChange}
          />
        </header>
        <div>
          {data.shows.map((show) => (
            <ShowCard key={show.imdbID} show={show} />
          ))}
        </div>
      </div>
    );
  }
}
