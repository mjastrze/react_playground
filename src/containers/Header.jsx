import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes, { string, bool } from 'prop-types';
import { connect } from 'react-redux';

import { setSearchTerm } from './actionCreators';

// //w nawiasach żeby interpreter nie brał jako ciało funkcji
const mapStateToProps = (state) => ({ searchTerm: state.searchTerm });
const mapDispatchToProps = (dispatch) => ({
  handleSearchTermChange(event) {
    dispatch(setSearchTerm(event.target.value));
  },
});

const getUtilSpace = (
  showSearch = false,
  searchTerm = '',
  handleSearchTermChange = () => {},
) =>
  showSearch ? (
    <input
      value={searchTerm}
      onChange={handleSearchTermChange}
      type="text"
      placeholder="Search"
    />
  ) : (
    <h2>
      <Link to="/search">Back</Link>
    </h2>
  );

const HeaderInternal = ({ showSearch, searchTerm, handleSearchTermChange }) => (
  <header>
    <Link to="/">
      <h1>video</h1>
    </Link>
    {getUtilSpace(showSearch, searchTerm, handleSearchTermChange)}
  </header>
);

export const Header = connect(
  mapStateToProps,
  mapDispatchToProps,
)(HeaderInternal);

Header.defaultProps = {
  showSearch: false,
  handleSearchTermChange: () => {},
  searchTerm: '',
};

Header.propTypes = {
  showSearch: bool,
  handleSearchTermChange: PropTypes.func,
  searchTerm: string,
};
