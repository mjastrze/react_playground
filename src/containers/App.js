import React from 'react';
import { hot } from 'react-hot-loader/root';
import { BrowserRouter, Route, Switch } from 'react-router-dom';

import { Landing } from './Landing';
import { Search } from './Search';

import '../public/style.css';

const FourOhFour = () => <h1>404</h1>;

const App = () => (
  <div className="app">
    <BrowserRouter>
      <Switch>
        <Route exact path="/" component={Landing} />
        <Route path="/search" component={Search} />
        <Route component={FourOhFour} />
      </Switch>
    </BrowserRouter>
  </div>
);

export default hot(App);
