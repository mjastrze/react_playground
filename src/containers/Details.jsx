import React from 'react';
import { connect } from 'react-redux';
import { shape, string } from 'prop-types';
import { Header } from './Header';
import { Spinner } from './Spinner';

import { getAPIDetails } from './actionCreators';

const mapStateToProps = (state, ownProps) => ({
  rating:
    state.apiData[ownProps.show.imdbID] &&
    state.apiData[ownProps.show.imdbID].rating,
});

const mapDispatchToProps = (dispatch, ownProps) => ({
  getAPIData() {
    dispatch(getAPIDetails(ownProps.show.imdbID));
  },
});

export const Details = connect(
  mapStateToProps,
  mapDispatchToProps,
)(
  class Details extends React.Component {
    componentDidMount() {
      const { rating, getAPIData } = this.props;

      if (!rating) {
        getAPIData();
      }
    }

    render() {
      const {
        show: { title, description, year, poster, trailer },
        rating,
      } = this.props;
      const ratingComponent = rating ? <h3>{rating}</h3> : <Spinner />;
      return (
        <div className="details">
          <Header />
          <section>
            <h1>{title}</h1>
            <h2>({year})</h2>
            {ratingComponent}
            <img
              src={`/public/img/posters/${poster}`}
              alt={`Poster for ${title}`}
            />
            <p>{description}</p>
          </section>
          <div>
            <iframe
              src={`https://www.youtube-nocookie.com/embed/${trailer}?rel=0&amp;showinfo=0`}
              title={`Trailer for ${title}`}
              frameBorder="0"
              allowFullScreen
            />
          </div>
        </div>
      );
    }
  },
);

Details.propTypes = {
  show: shape({
    poster: string.isRequired,
    title: string.isRequired,
    year: string.isRequired,
    description: string.isRequired,
    imdbID: string.isRequired,
    trailer: string.isRequired,
  }).isRequired,
};
