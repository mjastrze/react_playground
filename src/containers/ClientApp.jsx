import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, Switch } from 'react-router-dom';
import { Landing } from './Landing';
import { Search } from './Search';
import { Details } from './Details';
import preload from '../data.json';

import store from './store';
import '../public/style.css';

const FourOhFour = () => <h1>404</h1>;

export const App = () => (
  <div className="app">
    <BrowserRouter>
      <Provider store={store}>
        <Switch>
          <Route exact path="/" component={Landing} />
          <Route
            path="/search"
            component={(props) => <Search shows={preload.shows} {...props} />}
          />
          <Route
            path="/details/:id"
            component={(props) => (
              <Details
                show={preload.shows.find(
                  (show) => props.match.params.id === show.imdbID,
                )}
                {...props}
              />
            )}
          />
          <Route component={FourOhFour} />
        </Switch>
      </Provider>
    </BrowserRouter>
  </div>
);
