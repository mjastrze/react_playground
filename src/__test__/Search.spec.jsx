import React from 'react';
// import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { Search } from '../containers/Search';
import { ShowCard } from '../containers/ShowCard';
import data from '../data.json';

test('Search renders correctly', () => {
  // const component = renderer.create(<Search />);
  // const tree = component.toJSON();
  // expect(tree).toMatchSnapshot();

  const component = shallow(<Search />);
  expect(component).toMatchSnapshot();
});

test('Search should render correct amount of shows', () => {
  const component = shallow(<Search />);
  expect(component.find(ShowCard).length).toEqual(data.shows.length);
});

test('Search should render correct amount of shows based on search terms', () => {
  const searchWord = 'black';
  const component = shallow(<Search />);
  component.find('input').simulate('change', { target: { value: searchWord } });
  const showCount = data.shows.filter((show) =>
    `${show.title} ${show.description}`
      .toUpperCase()
      .includes(searchWord.toUpperCase()),
  ).length;
  expect(component.find(ShowCard).length).toEqual(showCount);
});
