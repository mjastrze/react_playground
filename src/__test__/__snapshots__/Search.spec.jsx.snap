// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders correctly 1`] = `
<div
  className="search"
>
  <header>
    <input
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value="dupa"
    />
  </header>
  <div />
</div>
`;
